File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/text/TemplateGroupsRule.java
package com.ririnto.sinon.harness.rules.text;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import com.ririnto.sinon.harness.rules.HarnessCheckRule;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;

/**
 * Rule that requires specific template groups to exist.
 */
public enum TemplateGroupsRule implements HarnessCheckRule {
    INSTANCE;

    private static final String CATEGORY = "templateGroups";

    @Override
    public String category() {
        return "templateGroups";
    }

    @Override
    public boolean applies(RuleContext ctx) {
        return ctx.manifest().isEnabled(CATEGORY);
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        final Path root = ctx.root();
        final JsonNode manifest = ctx.manifest().raw();
        final JsonNode catNode = manifest.get(CATEGORY);
        final String targetRoot = catNode.get("parameters").get("targetRoot").asString();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return HarnessCheckHelper.extractPaths(catNode.get("parameters").get("groups")).stream()
                .filter(g -> !HarnessCheckHelper.isSafeDirectory(root, root.resolve(targetRoot).resolve(g)))
                .map(g -> Finding.of(severity, CATEGORY, "missing template group: " + targetRoot + "/" + g))
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/fs/TemplateGroupsRule.kt
package com.ririnto.sinon.harness.rules.fs

import com.ririnto.sinon.harness.core.JsonAccess
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.rules.HarnessCheckRule
import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import com.ririnto.sinon.harness.core.Severity
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.isSymbolicLink

/**
 * Rule that requires template groups to exist as directories.
 */
object TemplateGroupsRule : HarnessCheckRule() {
    /**
     * Category key.
     */
    override val category: String = "templateGroups"

    override fun validate(ctx: RuleContext): Collection<Finding> = buildList {
        val catObj = ctx.manifest.categoryObject(category)
        val parametersObj = catObj?.get("parameters")?.jsonObject
        if (catObj != null && parametersObj != null) {
            for (group in ctx.manifest.stringArray(category, "groups")) {
                val p = ctx.root / JsonAccess.stringFromObject(parametersObj, "targetRoot") / group
                when {
                    p.isSymbolicLink() -> {
                        add(
                            Finding(
                                Severity.ERROR,
                                category,
                                "symlink directory is not allowed: ${JsonAccess.stringFromObject(parametersObj, "targetRoot")}/$group",
                            ),
                        )
                    }
                    !p.isDirectory() -> {
                        add(Finding(ctx.manifest.severityOf(category), category, "missing template group: ${JsonAccess.stringFromObject(parametersObj, "targetRoot")}/$group"))
                    }
                }
            }
        }
    }
}
